datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamResources.java

```diff
@@ -467,12 +472,16 @@ public ActionResult<Void> stop(@PathKeysParam PathKeys pathKeys,
           d.setStatus(DatastreamStatus.STOPPING);
           _store.updateDatastream(d.getName(), d, true);
           _store.deleteDatastreamNumTasks(d.getName());
+          // invoke post datastream state change action for recently stopped datastream
+          invokePostDSStateChangeAction(d);
         } else if (DatastreamStatus.STOPPING.equals(d.getStatus())) {
           // this check helps in preventing any datastream from being stuck in STOPPING state indefinitely
           LOG.warn("Datastream {} is already in {} state. Notifying leader to initiate transition", d,
               d.getStatus());
           _store.updateDatastream(d.getName(), d, true);
           _store.deleteDatastreamNumTasks(d.getName());
+          // invoke post datastream state change action for recently stopped datastream
+          invokePostDSStateChangeAction(d);
```

Comment:
The two invocations here for state change action are after setting the datastream status to STOPPING state, and we wait (in the polling loop below) in this stop rest call for the datastream to change to the STOPPED state before it is returned or, in the worst case, times out. 

So, instead of calling the invocation for the STOPPING state, should we call it after the poll loop so that the correct data stream status STOPPED is reflected? 

Also, for a much more consistent state change tracking, considering this invocation is lightweight, along with invoking these two invocations for the STOPPING state, we should have another invocation for the STOPPED state after the polling loop.

Reply:
Valid points

> So, instead of calling the invocation for the STOPPING state, should we call it after the poll loop so that the correct data stream status STOPPED is reflected?

The reason I added here and not after the polling, current logic written such a way that if one of the update fails it will throw the exception and return 5xx to client. So, the polling logic will not be invoked and we will not be able to update to DH. For example, if there are 5 datastreams need to be stopped, and if it causes failure at 3rd, we will throw the exception, so even though first 2 got stopped successfully, we will not have it reflected into DH.

>Also, for a much more consistent state change tracking, considering this invocation is lightweight, along with invoking these two invocations for the STOPPING state, we should have another invocation for the STOPPED state after the polling loop.

I like this idea about invoking this again. The connector can then decide what to do based on the datastream status. But, by the time connector gets the call from the 1st invocation, the datastream may have moved from stopping to stopped, so it may make it redundant, what do you think?

Reply:
Yes, we better update it in both states, Stopping and Stopped state. And if the connector is implementing it in separate threads, then there will be multiple updates to any external system where anyone might need to send it, which is what we want, I think.

That way, all the incremental updates to the DS are tracked properly.

Reply:
Ok, it would still not be tracked properly as we fail out the request if anyone of the update gives error. Also, when we perform action in the new method based on stopping call, the DS could be in stopped state already. However, I don't see a way to make it 100% accurate so by providing ability to call it for each action, STOPPING and STOPPED at least connector has a chance to act on it.

Reply:
Yep but you can send a new object to the connectors instead of sending the reference, so it could be tracked properly.

Reply:
Just for reference: what I mean, when we call postDSState...() on STOPPING, by the time the connector run the logic, it could be in the STOPPED state, however, the object that we have passed in has STOPPING. Not a big issue as we will anyway call postDSState...() again with status STOPPED.